File: rd-kt/rd-text/src/main/kotlin/com/jetbrains/rd/rdtext/intrinsics/RdTextChange.kt
package com.jetbrains.rd.rdtext.intrinsics

import com.jetbrains.rd.util.string.IPrintable
import com.jetbrains.rd.util.string.PrettyPrinter
import com.jetbrains.rd.util.string.printToString

enum class RdTextChangeKind {
    Insert,
    Remove,
    Replace,
    Reset,
    PromoteVersion,
    InsertLeftSide,
    InsertRightSide
}

class RdTextChange(val kind: RdTextChangeKind,
                   val startOffset: Int,
                   val old: String,
                   val new: String,
                   val fullTextLength: Int) : IPrintable {
    companion object {
        private fun escape(s: String): String = s.replace("\n", "\\n").replace("\r", "\\r")
    }

    override fun print(printer: PrettyPrinter) {
        printer.println("RdTextChange (")
        printer.indent {
            println("kind = $kind")
            println("startOffset = $startOffset")
            println("old = '${escape(old)}'")
            println("new = '${escape(new)}'")
            println("fullTextLength = $fullTextLength")
        }
        printer.println(")")
    }

    override fun toString(): String = this.printToString()
}

val RdTextChange.isNormalChange: Boolean get() = kind != RdTextChangeKind.Reset && kind != RdTextChangeKind.PromoteVersion

fun RdTextChange.delta(): Int {
    if (!isNormalChange) throw UnsupportedOperationException()
    return new.length - old.length
}

fun RdTextChange.reverse(): RdTextChange {
    val newKind = when (kind) {
        RdTextChangeKind.Insert -> RdTextChangeKind.Remove
        RdTextChangeKind.Remove -> RdTextChangeKind.Insert
        RdTextChangeKind.Replace -> RdTextChangeKind.Replace
        RdTextChangeKind.Reset -> throw UnsupportedOperationException()
        RdTextChangeKind.PromoteVersion -> throw UnsupportedOperationException()
        RdTextChangeKind.InsertLeftSide -> RdTextChangeKind.Remove
        RdTextChangeKind.InsertRightSide -> RdTextChangeKind.Remove
    }

    val textLength = if (fullTextLength == -1) -1 else fullTextLength - delta()
    return RdTextChange(newKind, startOffset, new, old, textLength)
}

fun RdTextChange.assertDocumentLength(current: Int) {
    if (!isNormalChange) return

    val actualLength = current + delta()
    val expectedLength = fullTextLength
    if (actualLength != expectedLength)
        throw IllegalStateException("Expected document length: $expectedLength, but actual: $actualLength")
}

File: rd-net/RdFramework/Text/Intrinsics/RdTextChange.cs
using System;
using JetBrains.Annotations;
using JetBrains.Rd.Base;
using JetBrains.Rd.Util;

namespace JetBrains.Rd.Text.Intrinsics
{
  public enum RdTextChangeKind
  {
    Insert,
    Remove,
    Replace,
    Reset,
    PromoteVersion,
    InsertLeftSide,
    InsertRightSide
  }

  public class RdTextChange : IPrintable
  {
    public readonly int FullTextLength;
    public readonly RdTextChangeKind Kind;

    public readonly string New;

    public readonly string Old;

    public readonly int StartOffset;

    public RdTextChange(RdTextChangeKind kind, int startOffset, string old, string @new, int fullTextLength) {
      Kind = kind;
      StartOffset = startOffset;
      Old = old;
      New = @new;
      FullTextLength = fullTextLength;
    }

    public void Print(PrettyPrinter printer) {
      printer.Println("RdTextChange (");
      using (printer.IndentCookie())
      {
        printer.Println($"Kind = {Kind}");
        printer.Println($"Start = {StartOffset}");
        printer.Println($"OldText = '{Escape(Old)}'");
        printer.Println($"NewText = '{Escape(New)}'");
        printer.Println($"FullTextLength = {FullTextLength}");
      }

      printer.Println(")");
    }

    public RdTextChange Reverse() {
      var newKind = ReverseKind(Kind);
      var fullTextLength = FullTextLength == -1 ? -1 : FullTextLength - this.GetDelta();
      return new RdTextChange(newKind, StartOffset, New, Old, fullTextLength);
    }

    private static RdTextChangeKind ReverseKind(RdTextChangeKind kind) {
      switch (kind)
      {
        case RdTextChangeKind.Insert: return RdTextChangeKind.Remove;
        case RdTextChangeKind.Remove: return RdTextChangeKind.Insert;
        case RdTextChangeKind.Replace: return RdTextChangeKind.Replace;
        case RdTextChangeKind.Reset: throw new InvalidOperationException("Reset change isn't invertible.");
        case RdTextChangeKind.PromoteVersion: throw new InvalidOperationException("PromoteVersion change isn't invertible.");
        case RdTextChangeKind.InsertLeftSide: return RdTextChangeKind.Remove;
        case RdTextChangeKind.InsertRightSide: return RdTextChangeKind.Remove;
        default: throw new ArgumentOutOfRangeException($"Unexpected Kind value: {kind}.");
      }
    }

    private static string Escape(string s) => s.Replace("\n", "\\n").Replace("\r", "\\r");

    public override string ToString() {
      var printer = new PrettyPrinter();
      Print(printer);
      return printer.ToString();
    }
  }
}